### DISCUSSION TITLE
Freezing tok2vec vs unfreezing tok2vec while updating pre-trained model

### DISCUSSION THREAD

**Discussion by umayerr** [explosion/spaCy#12518] [Help: Coding & Implementations] [answered]
I took a blank model and trained that (`modelA`) for `ner` with the vector of `en_core_web_lg` (for accuracy). After training it has only two components: `tok2vec` and `ner.` Now I have two scenarios below:

1. When I tried to update `modelA` with new dataset, I freeze `tok2vec` and I got `ValueError: [E954] The Tok2Vec listener did not receive any valid input from an upstream component`. However, I did not get this error when I updated a pre-trained model (e.g. `en_core_web_sm)` with a new dataset freezing all components except `ner` (`tok2vec` was in the freeze component list). I understand that `ner` has dependency on `tok2vec` because it needs vectorized tokens for recognition. If so, then how did the updating process work for `en_core_web_sm` where `tok2vec` was also in freeze component list?
2. When I tried updating `modelA` (keeping `tok2vec` unfreeze) I got `ValueError: Shape mismatch for blis.gemm: (0, 0), (1280, 768)`. For checking, I took another blank model and trained that (`testModel`) for `ner` without any vector (for efficiency) and I could successfully update the `testModel` with new dataset. Why are these models behaving differently while updating with new data? Is this because of the vector of `en_core_web_lg`? How can I resolve this error? 
Here is the code for updating model with new data.
```
# loading model and getting pipe
nlp = spacy.load('modelA/model-best/')
ner = nlp.get_pipe('ner')

# checking labels from training data and adding to the pipe if neccessary
existing_label = list(ner.labels)
label_in_train_data = []

for text, annotation in TRAIN_DATA:
    for entity in annotation.get("entities"):
        if entity[2] not in label_in_train_data:
            label_in_train_data.append(entity[2])
            
for label in label_in_train_data:
    if label not in existing_label:
        ner.add_label(label)

# blocking other pipes to be updated and starting training
affected_pipe = ['tok2vec', 'ner']
unaffected_pipe = [pipe for pipe in nlp.pipe_names if pipe not in affected_pipe]

with nlp.disable_pipes(*unaffected_pipe):
    optimizer = nlp.create_optimizer()
    for iter_ in range(10):
        random.shuffle(TRAIN_DATA)
        losses = {}
        for text, annotation in tqdm(TRAIN_DATA):
            example = []
            doc = nlp.make_doc(text)
            print(doc)
            print(doc.vector.shape)
            example.append(Example.from_dict(doc, annotation))
            nlp.update(example, drop=0.2, sgd=optimizer, losses=losses)
nlp.to_disk('newModel')
```

**Comment by svlandeg** [comment, marked as answer]
Hi!

>When I tried to update modelA with new dataset, I freeze tok2vec and I got ValueError: [E954] The Tok2Vec listener did not receive any valid input from an upstream component. 

If you have a `tok2vec` component in the pipeline that you want to freeze but you also want to continue training other components on top of it, then you should add this `tok2vec` component to the  [`annotating_components`](https://spacy.io/usage/training#annotating-components) section in your config:

```
[training]
annotating_components = ["tok2vec"]
```

This will basically ensure that the component runs during training but isn't backpropagated into - i.e. it won't update as long as it's (also) in the frozen components list.

> However, I did not get this error when I updated a pre-trained model (e.g. en_core_web_sm)

That's a good question with a somewhat complex answer but that might help you better understand what's going on. You can see the configuration of the models in the config file of `en_core_web_sm`. It has this pipeline:
```
pipeline = ["tok2vec","tagger","parser","senter","attribute_ruler","lemmatizer","ner"]
```
But the `tok2vec` component here in fact is only used by the `parser` and the `tagger`, who use it through the "listener" mechanism:
```
[components.tagger.model.tok2vec]
@architectures = "spacy.Tok2VecListener.v1"
```
In those cases (like in your original  case), when you're using the listener mechanism for the `tok2vec` layer, you can freeze it + put it in `annotating_components` as I explained.

The `ner` component of that specific pipeline however, has its own sort of "private" `tok2vec` model which is part of the NER model only:
```
[components.ner.model.tok2vec]
@architectures = "spacy.Tok2Vec.v2"
```
This means that if you freeze the whole pipeline except for the NER, you're actually still (re)training the internal `tok2vec` layer of that NER model - it's not frozen. You can only freeze full components, like the `tok2vec` component that is defined in `[components.tok2vec]`. 

More background information on those two different patterns can be found here: https://spacy.io/usage/embeddings-transformers; it's basically the difference between having a "shared" Tok2Vec layer or an independent one - the first is a separate component by itself, the second is an independent layer within the model of another component.

**Reply by umayerr** [reply]
Hi @svlandeg thank you for the resource.
In the config file I have noticed that `ner` component used shared `tok2vec` through the 'listener' mechanism like `tagger` and `parser`. I created the config file using CLI command and I did not change anything after generating it.
```
[components.ner.model.tok2vec]
@architectures = "spacy.Tok2VecListener.v1"

[components.tok2vec.model]
@architectures = "spacy.Tok2Vec.v2"
```
I am wondering why spaCy supplied `tok2vec` to `ner` component through the 'listener' mechanism for training a blank model? Because it passes `tok2vec` without 'listener' mechanism to `ner` component of a pre-trained model. So, if I want to train a model for `ner` from scratch as like the `ner` component of spaCy pre-trained model then I have to correct the config file manually, is that right?

**Reply by svlandeg** [reply]
Yes, that's right. The reason why the CLI command suggests the "listener" architecture is that this would be more generally applicable when someone wants to train several components on the same dataset and using the same embedding layer.

If you're only training an NER and don't need the listener architecture, you can have a look at `en_core_web_lg`'s configuration of the `ner` component and mimick that. However - I'm not sure you need to do that? You can also just use the listener architecture for one component - then you'll have a `tok2vec` and an `ner` in the pipeline and you should be able to do the tok2vec freezing + NER (re)training by having the `tok2vec` in the annotating components list.

(I know this can all get a bit confusing so feel free to ask more questions for clarification!)

FYI, we also have a command [`debug config`](https://spacy.io/api/cli#debug-config) that could be useful to you once you start editing the config files ;-)

**Reply by umayerr** [reply]
Thank you so much for your explanation. I think I am almost there. So, if I understand correctly, this is the same `tok2vec` with two different ways of implementation; `shared` and `independent`. The same configuration, I found, for both `shared` and `independent` `tok2vec` in the `en_core_web_lg`'s config file. 

Now I have just one confusion (maybe trivial) left from my very first post (question 2). If I simplify that question: I trained two blank models only for `ner` with `en_core_web_lg`'s vectors (for accuracy). The only difference between that two models is one used `shared` `tok2vec` and another used `independent` one. After that when I tried to update that two models with new data, the model with `independent` `tok2vec` updated successfully. But the model with `shared` `tok2vec` did not update. It ended up with the `ValueError: Shape mismatch for blis.gemm: (0, 0), (1280, 768)`. 

Please note that updating models with new data works fine (for both `shared` and `independent` approaches) if I do not use any existing vectors while training the blank models.

**Reply by svlandeg** [reply]
> So, if I understand correctly, this is the same tok2vec with two different ways of implementation; shared and independent.

That's right. Think of it like this: the `tok2vec` model is just a bunch of neural network layers, bundled together as a Thinc model (Thinc is our internal ML library). Such a Thinc model can just be used as another "layer" or subnetwork in another Thinc model - that's what's happening for the NER of the `en_core_web_lg` model, so the `tok2vec` model is internal/independent. In contrast, the other approach is wrapping the `tok2vec` Thinc model into an actual spaCy `tok2vec` component which shows up as a separate component in `nlp.pipeline`. This can then be shared/listened to by various other components. At the core, the Thinc models are the same, they're just represented differently. 

(let me write a separate response for your other question)

**Reply by svlandeg** [reply]
> But the model with shared tok2vec did not update. It ended up with the ValueError: Shape mismatch for blis.gemm: (0, 0), (1280, 768).

I can't immediately reproduce this. Can you share the config file you used, as well as the code/commands you run as a sort of minimal reproducible example that lets me reproduce the error and investigate what might be the cause of this?

**Reply by umayerr** [reply]
Sure! Here it is,
**Training blank model:**
Steps I followed:
1. Creating `config.cfg` file. Here is the [config](https://drive.google.com/file/d/1vz7cQg8SX09Qoi6dNrn0gQKMRtrimNJG/view?usp=sharing) file. I used `en_core_sci_md`'s vectors to train the blank model. The model `en_core_sci_md` is provided by Scispacy for recognizing scientific data. 
2. Creating [train.spacy](https://drive.google.com/file/d/1NGHXITdIjJy-GUr8nsMiI00a4W5W6-XO/view?usp=sharing) and [dev.spacy](https://drive.google.com/file/d/188QFClCRAH4kCpC3NdGcN_dh9iHwQO5e/view?usp=sharing). I have reduced both files with 5 examples each.
3. Training model with `python -m spacy train config.cfg --output ./output --paths.train ./train.spacy --paths.dev ./dev.spacy` command. 
4. Based on the configuration mentioned above, I have also added a small TRAINED_MODEL (trained with 5 examples) [here](https://drive.google.com/drive/folders/1Q4hqbyj0PUJiO0deSz2IS2Ql2ir30Xmj?usp=sharing). For the simplicity I used same data (`train.spacy`) for training and development of the model. I got the error when I tried to update this TRAINED_MODEL following the steps described below.

**Updating the trained model with new data:**
Steps I followed:
1. Converting [JSON examples](https://drive.google.com/file/d/1RST4LezGfSJPVYz6E8JcYb25Cs6MoiiF/view?usp=sharing) to spacy training data format. This is the [code](https://drive.google.com/file/d/1naPEtc3YPPUGMxNxaOTklj7MQEaxA0Qt/view?usp=sharing) for converting JSON to spacy data. 
2. Updating the model with new data. This is the [code](https://drive.google.com/file/d/1SeXopU2vpgYujrjaxgIHIfXm4RaRZx5N/view?usp=sharing).

**Reply by svlandeg** [reply]
Hi! Sorry for the delayed response - I lost track of this due to a release deadline and some holidays. I'll look into this in more detail later this week (if it's still an issue?)

**Reply by svlandeg** [reply]
Hi @ureza!

Thanks for providing the files and detailed instructions, that's always super helpful. Unfortunately, I still haven't been able to reproduce this. Your config looks fine, so does the script, and on my end I can get it to run on new data while adding new labels to the `ner` on the fly, as expected.

Can you try a few things for me, each independently and see whether perhaps that fixes the issues on your end? It'll help us narrow things down.

1. Replace `en_core_sci_md` with `en_core_web_md` in your config file, then run the whole flow again (training + retraining)
2. Try running everything on spaCy 3.5.1 instead (if I'm not mistaken, Scispacy is currently limited to <3.5, so you'll want to use `en_core_web_md`) 
3. Instead of the JSON conversion, try with some data defined in-code, e.g.

```
    TRAIN_DATA = [
        (
            "Russ Cochran captured his first major title with his son as caddie.",
            {"entities": [(0, 12, "PERSON")]},
        ),
        (
            "Russ Cochran his reprints include EC Comics.",
            {"entities": [(0, 12, "PERSON"), (34, 43, "ART")]},
        ),
    ]
```

**Reply by umayerr** [reply]
Hi @svlandeg!
Apologies for the delayed response. I'm aware of the recent news regarding the release of LLMs, and I hope it was a successful endeavor. On a positive note, I managed to identify the issue and successfully resolved it, allowing the entire flow to work seamlessly. The problem originated from the training data, specifically an overlooked empty annotation (like below). I appreciate your continued support and patience throughout this process. I sincerely apologize for overlooking such a trivial aspect in the training data. Thank you once again for helping me.
```
    [
      "",
      {
        "entities": []
      }
    ]
```

**Reply by svlandeg** [reply]
Hi @ureza , no need to apologize at all! I'm happy to hear you got the issue resolved on your end 🎉
